datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -121,58 +291,220 @@ where
+                )))
+            }
+            other => {
+                internal_err!("Unsupported data type {other:?} for function {name}")
+            }
+        },
+        // if the first argument is a scalar utf8 all arguments are expected to be scalar utf8
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) | ScalarValue::LargeUtf8(a) => {
+                let mut val: Option<Result<ColumnarValue>> = None;
+                let mut err: Option<DataFusionError> = None;
+
+                match a {
+                    Some(a) => {
+                        // enumerate all the values finding the first one that returns an Ok result
+                        for (pos, v) in args.iter().enumerate().skip(1) {
+                            if let ColumnarValue::Scalar(s) = v {
+                                if let ScalarValue::Utf8(x) | ScalarValue::LargeUtf8(x) =
+                                    s
+                                {
+                                    if let Some(s) = x {
+                                        match op(a.as_str(), s.as_str()) {
+                                            Ok(r) => {
+                                                val = Some(Ok(ColumnarValue::Scalar(
+                                                    S::scalar(Some(op2(r))),
+                                                )));
+                                                break;
+                                            }
+                                            Err(e) => {
+                                                err = Some(e);
+                                            }
+                                        }
+                                    }
+                                } else {
+                                    return internal_err!("Unsupported data type {s:?} for function {name}, arg # {pos}");
+                                }
+                            } else {
+                                return internal_err!("Unsupported data type {v:?} for function {name}, arg # {pos}");
+                            }
+                        }
+                    }
+                    None => (),
+                }
+
+                if let Some(v) = val {
+                    v
+                } else {
+                    Err(err.unwrap())
+                }
+            }
+            other => {
+                internal_err!("Unsupported data type {other:?} for function {name}")
+            }
+        },
+    }
+}
+
 /// Calls string_to_timestamp_nanos and converts the error type
 fn string_to_timestamp_nanos_shim(s: &str) -> Result<i64> {
     string_to_timestamp_nanos(s).map_err(|e| e.into())
 }
 
+/// Calls string_to_timestamp_nanos_formatted and converts the error type
+fn string_to_timestamp_nanos_with_format_shim(s: &str, f: &str) -> Result<i64> {
+    string_to_timestamp_nanos_formatted(s, f)
```

Comment:
No need for shim here it seems, as there's no error conversion happening?

Also, as drive-by, `string_to_timestamp_nanos_shim` above can be simplified to return `Ok(string_to_timestamp_nanos(s)?)`, which is a bit more idiomatic.

Reply:
Ah, good catch. It used to when I had this code in arrow_rs, I lifted it out of there and forgot to refactor this.